lib/gateway/build-gateway.js

```diff
@@ -508,12 +512,25 @@ async function buildGateway (serviceMap, gatewayOpts, app, lruGatewayResolvers)
     serviceMap,
     entityResolversFactory: factory,
     pollingInterval: gatewayOpts.pollingInterval,
+    serviceFn: gatewayOpts.services instanceof Function ? gatewayOpts.services : undefined,
     async refresh (isRetry) {
       const failedMandatoryServices = []
       if (this._serviceSDLs === undefined) {
         this._serviceSDLs = serviceSDLs.join(' ')
       }
 
+      if (this.serviceFn) {
+        const newServices = await this.serviceFn()
+        const oldServices = Object.keys(serviceMap)
+        const addedServices = newServices.filter(({ name }) => !oldServices.includes(name))
+        const deletedServices = oldServices.filter(name => !newServices.find(service => service.name === name))
+        for (const name of deletedServices) {
+          await serviceMap[name].close()
```

Comment:
This might be problematic and take a long time. It might be the cause of the actual instability.

Maybe try:

```suggestion
          serviceMap[name].close().catch(() => {})
```

Reply:
I had to add an ignore for the coverage of the catch branch. Let's see if it passes now